src/cargo/ops/cargo_uninstall.rs

```diff
@@ -74,80 +75,74 @@ pub fn uninstall_one(
     bins: &[String],
     config: &Config,
 ) -> CargoResult<()> {
-    let crate_metadata = metadata(config, root)?;
-    let metadata = read_crate_list(&crate_metadata)?;
-    let pkgid = PackageIdSpec::query_str(spec, metadata.v1().keys().cloned())?;
-    uninstall_pkgid(&crate_metadata, metadata, pkgid, bins, config)
+    let tracker = InstallTracker::load(config, root)?;
+    let all_pkgs = tracker.all_installed_bins().map(|(pkg_id, _set)| *pkg_id);
+    let pkgid = PackageIdSpec::query_str(spec, all_pkgs)?;
+    uninstall_pkgid(root, tracker, pkgid, bins, config)
 }
 
 fn uninstall_cwd(root: &Filesystem, bins: &[String], config: &Config) -> CargoResult<()> {
-    let crate_metadata = metadata(config, root)?;
-    let metadata = read_crate_list(&crate_metadata)?;
+    let tracker = InstallTracker::load(config, root)?;
     let source_id = SourceId::for_path(config.cwd())?;
     let src = path_source(source_id, config)?;
     let (pkg, _source) = select_pkg(src, None, None, config, true, &mut |path| {
         path.read_packages()
     })?;
     let pkgid = pkg.package_id();
-    uninstall_pkgid(&crate_metadata, metadata, pkgid, bins, config)
+    uninstall_pkgid(root, tracker, pkgid, bins, config)
 }
 
 fn uninstall_pkgid(
-    crate_metadata: &FileLock,
-    mut metadata: CrateListingV1,
+    root: &Filesystem,
+    mut tracker: InstallTracker,
     pkgid: PackageId,
     bins: &[String],
     config: &Config,
 ) -> CargoResult<()> {
     let mut to_remove = Vec::new();
-    {
-        let mut installed = match metadata.v1_mut().entry(pkgid) {
-            Entry::Occupied(e) => e,
-            Entry::Vacant(..) => failure::bail!("package `{}` is not installed", pkgid),
-        };
-
-        let dst = crate_metadata.parent().join("bin");
-        for bin in installed.get() {
-            let bin = dst.join(bin);
-            if fs::metadata(&bin).is_err() {
-                failure::bail!(
-                    "corrupt metadata, `{}` does not exist when it should",
-                    bin.display()
-                )
-            }
-        }
-
-        let bins = bins
-            .iter()
-            .map(|s| {
-                if s.ends_with(env::consts::EXE_SUFFIX) {
-                    s.to_string()
-                } else {
-                    format!("{}{}", s, env::consts::EXE_SUFFIX)
-                }
-            })
-            .collect::<Vec<_>>();
+    let installed = match tracker.installed_bins(pkgid) {
+        Some(bins) => bins.clone(),
+        None => bail!("package `{}` is not installed", pkgid),
+    };
 
-        for bin in bins.iter() {
-            if !installed.get().contains(bin) {
-                failure::bail!("binary `{}` not installed as part of `{}`", bin, pkgid)
-            }
+    let dst = root.join("bin").into_path_unlocked();
```

Comment:
To avoid extra uses of `into_path_unlocked`, could the bin path be learned from `InstallTracker` which internally holds locks?

Reply:
I'm not sure I understand. Do you mean something like this?

```rust
    pub fn bin_dir(&self) -> PathBuf {
        self.v1_lock.parent().join("bin")
    }
```

Reply:
Something like that yeah, but actually in retrospect I'm not sure if this is a great idea due to `--no-track`. In general `into_path_unlocked` is a code smell we want to avoid and is akin to `unsafe` sorta. In that sense we'd ideally thoroughly audit any usages of `into_path_unlocked` regularly (hah, like that ever will happen) and have sufficient comments indicating why it's justified to not require file locking at a particular time.

In the spirit of that I was largely hoping to minimize the number of calls to `into_path_unlocked`, but `InstallTracker` here isn't always available. Maybe though a method could be added to `InstallTracker` and then there'd be one-ish invocation of `into_path_unlocked` for `--no-track`, and a comment saying that if you say `--no-track` it's up to someone else to handle concurrent installs?

Reply:
I thought about this for a while.  I'm reluctant to add `bin_dir` if it is used in one place but no other.  I also don't see a clear way to reduce the `into_path_unlocked`.  The `dst` for install is used in a large number of places.  We could create a new abstraction (similar to `Layout`) which has a temporary lock file in `$CARGO_INSTALL_ROOT` which could wrap all the usage in a safer way.  In this case, I don't see it significantly improving things.

FWIW, I'm not yet convinced on the utility of `--no-track`.  If that is removed, something like `bin_dir` would work better.

Reply:
Ok sounds reasonable to me!